tests/transports/utils.nim

```diff
@@ -131,3 +134,70 @@ proc extractPort*(ma: MultiAddress): int =
   let portBytes = ma[codec].tryGet().protoArgument().tryGet()
   let port = int(fromBytesBE(uint16, portBytes))
   port
+
+template noException*(stream: Connection, body) =
+  try:
+    body
+  except CatchableError as exc:
+    raiseAssert "should not fail: " & exc.msg
+  finally:
+    await stream.close()
+
+proc serverHandlerSingleStream*(
+    server: Transport,
+    streamProvider: StreamProvider,
+    handler: proc(stream: Connection) {.async: (raises: []).},
+) {.async: (raises: []).} =
+  try:
+    let conn = await server.accept()
+    let muxer = streamProvider(server, conn)
+    muxer.streamHandler = handler
+
+    let muxerTask = muxer.handle()
+
+    await muxerTask
+    await muxer.close()
+    await conn.close()
+  except CatchableError as exc:
+    raiseAssert "should not fail: " & exc.msg
+
+proc clientRunSingleStream*(
+    server: Transport,
+    transportProvider: TransportProvider,
+    streamProvider: StreamProvider,
+    handler: proc(stream: Connection) {.async.},
+) {.async: (raises: []).} =
+  try:
+    let client = transportProvider()
+    let conn = await client.dial("", server.addrs[0])
+    let muxer = streamProvider(client, conn)
+    let muxerTask = muxer.handle()
+    asyncSpawn muxerTask
+
+    let stream = await muxer.newStream()
+    await handler(stream)
+
+    await stream.close()
+    await muxer.close()
+    await conn.close()
+    await muxerTask
+  except CatchableError as exc:
+    raiseAssert "should not fail: " & exc.msg
+
+proc runSingleStreamScenario*(
+    multiAddress: seq[MultiAddress],
+    transportProvider: TransportProvider,
+    streamProvider: StreamProvider,
+    serverStreamHandler: proc(stream: Connection) {.async: (raises: []).},
+    clientStreamHandler: proc(stream: Connection) {.async.},
```

Comment:
would be nice of these two are same types, for consistency. that is clientStreamHandler to have same raises as serverStreamHandler